### PR TITLE
Trivially Citable Citations: Base Class

### LINKED ISSUE #26160: Trivially exportable citations: Citation types and conversion functions
Re #26159.

Refer to design document [here](https://github.com/mantidproject/documents/pull/67).

Implement:
- `Citation` class
- Citation type information `struct`s
- Functions to create `Citation` from each type `struct`

By the end of this you should be able to do something like:
```cpp
auto const savu = Citation(PaperCitation{
  {"Wadeson, N", "Basham, M"}, "Savu: A Python-based, MPI Framework for Simultaneous Processing of Multiple, N-dimensional, Large Tomography Datasets", "", 2016, ""}
);
```

### PR DESCRIPTION
**Pull request by Pasarus**
**Description of work.**
The aim was to implement the base class of this design documentation:
https://github.com/mantidproject/documents/pull/67

**To test:**
Check that this works in the Script windows of MantidPlot and Workbench:
```python
from mantid.api import Citation 

cite = Citation("doi", "bibtex", "endnote", "url", "description")
print(cite.doi())
print(cite.bibtex())
print(cite.endnote())
print(cite.url())
print(cite.description())
```

Fixes #26160 . 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by DanNixon** [comment]
There's no need for release notes until the entire feature is complete.

File: Framework/API/inc/MantidAPI/Citation.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  const std::string &endnote() const;
+
+  void loadNexus(::NeXus::File *file, const std::string &group);
+  void saveNexus(::NeXus::File *file, const std::string &group);
+
+private:
+  // Optional
+  std::string m_description;
+
+  // If this is set doi, bibtex or endote are provided then url must be.
```
Comment (DanNixon) [inline Framework/API/inc/MantidAPI/Citation.h]: Could this be rephrased? It's unclear what this means at the moment.

File: Framework/API/src/Citation.cpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  and Endnote can be generated from DOIs)
+    - if none of doi, bibtex or endnote are provided, url must be provided
+  (there must be something there, even if this isn't citable a URL is better
+  than nothing)
+ */
+
+Citation::Citation(const std::string &doi, const std::string &bibtex,
+                   const std::string &endnote, const std::string &url,
+                   const std::string &description) {
+  if (doi == "" && bibtex == "" && endnote == "" && url == "" &&
```
Comment (DanNixon) [inline Framework/API/src/Citation.cpp]: Use `std::string::empty` to check for empty strings.

File: Framework/API/src/Citation.cpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+  if (doi != "" && (bibtex == "" || endnote == "" || url == ""))
+    throw std::invalid_argument(
+        "If doi is provided then url, bibtex and endnote must be");
+
+  if (doi == "" && bibtex == "" && endnote == "")
+    if (url == "")
+      throw std::invalid_argument(
+          "If none of doi, bibtex, or endnote is provided, then url must be");
+
+  m_doi = doi;
```
Comment (DanNixon) [inline Framework/API/src/Citation.cpp]: This can be done in the member initializer list.

File: Framework/API/src/Citation.cpp [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+const std::string &Citation::bibtex() const { return m_bibtex; }
+const std::string &Citation::endnote() const { return m_endnote; }
+
+void Citation::loadNexus(::NeXus::File *file, const std::string &group) {
+  file->openGroup(group, "NXCite");
+  file->readData("url", m_url);
+  file->readData("description", m_description);
+  file->readData("doi", m_doi);
+  file->readData("endnote", m_endnote);
+  // Since it remove the whitespace and it intends to restore it's state
```
Comment (DanNixon) [inline Framework/API/src/Citation.cpp]: What is the newline for? Why is this here for EndNote but not BibTex?

File: Framework/API/inc/MantidAPI/CitationConstructorHelpers.h [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+
+#include "DllConfig.h"
+
+#include <boost/optional.hpp>
+#include <vector>
+
+namespace Mantid {
+namespace API {
+
+struct MANTID_API_DLL BaseCitation {
```
Comment (DanNixon) [inline Framework/API/inc/MantidAPI/CitationConstructorHelpers.h]: Why did you go with this approach rather then free functions that return `Citation`?

This approach has some rather annoying usability issues:
- it makes constructing a `Citation` more work (first having to create a helper object then create a `Citation` from that
- no automatic population of URL or DOI fields

**Review by DanNixon — CHANGES_REQUESTED** [review]

File: Framework/API/inc/MantidAPI/CitationConstructorHelpers.h (hunk 5 above)
Comment (Pasarus) [reply Framework/API/inc/MantidAPI/CitationConstructorHelpers.h]: - You do not have to create a helper object first, it is 1 of 3 options. It said in the design document to add helper structs for easier construction of citations.
- All base citations should pass URL and or DOI into the respective Citation object.

These don't actually seem like problems.

I went with this approach as this way I don't need to rewrite 14 similar free functions, and instead I can utilise some basic polymorphism for easier implementation.

File: Framework/API/src/Citation.cpp (hunk 4 above)
Comment (Pasarus) [reply Framework/API/src/Citation.cpp]: Endnote will be initialised in an RSI file format style which will always need a new line there. Whilst BibTex will not.

File: Framework/API/inc/MantidAPI/CitationConstructorHelpers.h (hunk 5 above)
Comment (DanNixon) [reply Framework/API/inc/MantidAPI/CitationConstructorHelpers.h]: > These don't actually seem like problems.

If I have to specify a URL in up to three places that is a problem.

Can you show a complete example of creating a citation for a paper (i.e. on that has all four entries)?

File: Framework/API/src/Citation.cpp (hunk 4 above)
Comment (DanNixon) [reply Framework/API/src/Citation.cpp]: Not entirely sure what you mean here, at least in BibTex you have one entry per line, s each line will always end in a newline.

Regardless that is something we worry about when we export the citations, not here.

Comment (Pasarus) [reply Framework/API/src/Citation.cpp]: It was present before being loaded so it is being put back. That is the point of this.

File: Framework/API/inc/MantidAPI/CitationConstructorHelpers.h (hunk 5 above)
Comment (Pasarus) [reply Framework/API/inc/MantidAPI/CitationConstructorHelpers.h]:
```python
cite = mantid.api.citation(mantid.api.ArticleCitation(["author1"], "title", "journal", "year", "volume", "number", "pages", "month", "description", "doi", "url"))
url = cite.url()
```

```cpp
Mantid::API::Citation cite(Mantid::API::ArticleCitation({"author1"}, "title", "journal", "year", "volume", "number", "pages", "month", "description", "doi", "url")));
auto url  = cite.url();
```

Something like that will work.

Comment (Pasarus) [reply Framework/API/inc/MantidAPI/CitationConstructorHelpers.h]: url in these cases will be "url".

Comment (DanNixon) [reply Framework/API/inc/MantidAPI/CitationConstructorHelpers.h]: Ok, I see now.

However I feel this solution is overly complex and the polymorphism isn't actually gaining you anything (looking at the actual implementations of `toBibtex` and `toEndnote` backs this up).

Earlier we had discussed these conversion functions just being free functions that take the same parameters and return a `Citation`. What was limiting about this that made you implement the solution you present here?

Comment (Pasarus) [reply Framework/API/inc/MantidAPI/CitationConstructorHelpers.h]: - There is very little cost to this
- As far as I am concerned 3 overloaded citation constructors is easier to understand than have 3 random functions.
- This follows object orientated design principles which Mantid does in the majority of its code

Comment (DanNixon) [reply Framework/API/inc/MantidAPI/CitationConstructorHelpers.h]: - `CitationConstructorHelpers.h` and `CitationConstructorHelpers.cpp` contain more code than they would if you had used free functions in the way we had previously discussed
- OOP has a place, don't feel you have to use it just because Mantid already does in places
- OOP does not allow for much code reuse here and trying to shoehorn it in increases complexity

I say we leave this PR for now until we can discuss it in person next week.

Comment (Pasarus) [reply Framework/API/inc/MantidAPI/CitationConstructorHelpers.h]: Can you please give me an example of why a free function implementation would be better?

File: Framework/API/inc/MantidAPI/Citation.h [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+#include <string>
+#include <vector>
+
+#include <nexus/NeXusFile.hpp>
+
+namespace Mantid {
+namespace API {
+
+class MANTID_API_DLL Citation {
```
Comment (Pasarus) [inline Framework/API/inc/MantidAPI/Citation.h]:
```suggestion
/**
add example citation construction
*/
class MANTID_API_DLL Citation {
```

File: Framework/API/inc/MantidAPI/Citation.h [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+#include <string>
+#include <vector>
+
+#include <nexus/NeXusFile.hpp>
+
+namespace Mantid {
+namespace API {
+
+/**
```
Comment (Pasarus) [inline Framework/API/inc/MantidAPI/Citation.h]:
```suggestion
/*
```

File: Framework/API/inc/MantidAPI/Citation.h [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+#include <string>
+#include <vector>
+
+#include <nexus/NeXusFile.hpp>
+
+namespace Mantid {
+namespace API {
+
+/*
+ * To construct a citation class give it properly formatted bibtex and endnote
```
Comment (DanNixon) [inline Framework/API/inc/MantidAPI/Citation.h]: Can these be moved to Doxygen comments on the relevant constructors?

File: Framework/API/inc/MantidAPI/Citation.h [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+  // Needed for future Set constructiom
+  bool operator==(const Citation &rhs) const;
+
+  const std::string &description() const;
+  const std::string &url() const;
+  const std::string &doi() const;
+  const std::string &bibtex() const;
+  const std::string &endnote() const;
+
+  void loadNexus(::NeXus::File *file, const std::string &group);
```
Comment (DanNixon) [inline Framework/API/inc/MantidAPI/Citation.h]: Why do we have a constructor to create from NeXus and a `loadNexus` function?

We only need one way of doing this.

File: Framework/API/src/Citation.cpp [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+#include "MantidAPI/Citation.h"
+
+#include <nexus/NeXusFile.hpp>
+
+namespace Mantid {
+namespace API {
+
+/**
+  This function is designed for construction and validation for loading, however
```
Comment (DanNixon) [inline Framework/API/src/Citation.cpp]: What is this comment referencing? The constructor?

Also what is `getCitation`?.

**Review by DanNixon — CHANGES_REQUESTED** [review]

File: Framework/API/inc/MantidAPI/Citation.h (hunk 9 above)
Comment (Pasarus) [reply Framework/API/inc/MantidAPI/Citation.h]: The Constructor is so there is no way to construct a citation manually, the loadNexus function is called directly, I left it under public API so someone could load the values to the Citation again.

File: Framework/API/src/Citation.cpp (hunk 10 above)
Comment (Pasarus) [reply Framework/API/src/Citation.cpp]: Legacy comment, I shall clean it up.

File: Framework/API/inc/MantidAPI/Citation.h (hunk 9 above)
Comment (Pasarus) [reply Framework/API/inc/MantidAPI/Citation.h]: Also it is mainly aimed at usability when the loading from nexus files is fully implemented

Comment (DanNixon) [reply Framework/API/inc/MantidAPI/Citation.h]: :+1: In that case remove the `loadNexus` function.

Comment (Pasarus) [reply Framework/API/inc/MantidAPI/Citation.h]: I think private is best here.

File: Framework/API/inc/MantidAPI/Citation.h (hunk 8 above)
Comment (Pasarus) [reply Framework/API/inc/MantidAPI/Citation.h]: I will leave it as a normal comment as Doxygen keeps trying to interpret the endnote string as commands

Comment (DanNixon) [reply Framework/API/inc/MantidAPI/Citation.h]: Just remove the examples and link to an online reference.

**Review by DanNixon — APPROVED** [review]
